sharedbuffers/mapped_struct.py

```diff
@@ -1001,6 +1001,166 @@ def __str__(self):
+class proxied_frozenset(object):
+
+    cython.declare(objlist = object)
+
+    def __init__(self, objlist):
+        self.objlist = objlist
+
+    @classmethod
+    def pack_into(cls, obj, buf, offs, idmap = None, implicit_offs = 0):
+        return mapped_frozenset.pack_into(obj, buf, offs, idmap, implicit_offs, sort=True)
+
+    @classmethod
+    @cython.locals(
+        i=int, j=int, offs=cython.longlong,
+        pybuf='Py_buffer', pbuf='const unsigned char *', b=cython.uchar)
+    def unpack_from(cls, buf, offs, idmap = None):
+        buf = _likerobuffer(buf)
+        if cython.compiled:
+            PyObject_GetBuffer(buf, cython.address(pybuf), PyBUF_SIMPLE)
+            pbuf = cython.cast(cython.p_uchar, pybuf.buf)
+            if offs >= pybuf.len:
+                PyBuffer_Release(cython.address(pybuf))
+                raise IndexError("Offset out of range")
+        else:
+            pbuf = buf
+        try:
+            if pbuf[offs] == 'm':
+                # inline bitmap
+                if cython.compiled and offs+7 >= pybuf.len:
+                    raise IndexError("Object spans beyond buffer end")
+                rv = []
+                for i in xrange(7):
+                    b = ord(pbuf[offs+1+i])
+                    if b:
+                        for j in xrange(8):
+                            if b & (1<<j):
+                                rv.append(i*8+j)
+                return frozenset(rv)
+            else:
+                return proxied_frozenset(proxied_list(buf, offs, idmap))
+        finally:
+            if cython.compiled:
+                if type(buf) is buffer:
+                    PyBuffer_Release(cython.address(pybuf))
+
+    def copy(self):
+        return self
+
+    def __len__(self):
+        return len(self.objlist)
+
+    @cython.locals(i1=cython.Py_ssize_t, i2=cython.Py_ssize_t,
+        cnt=cython.Py_ssize_t, step=cython.Py_ssize_t, lst='proxied_list')
+    def __contains__(self, elem):
+        i1 = 0
+        i2 = len(self.objlist)
+        cnt = i2 - i1
+        lst = self.objlist
+
+        while cnt > 0:
+            step = cnt >> 1
+            val = lst._getitem(step)
+            if val == elem:
+                return True
+            elif val < elem:
+                cnt -= step + 1
+            else:
+                cnt = step
+        return False
+
+    def __iter__(self):
+        return iter(self.objlist)
+
+    def union(self, *seqs):
+        return frozenset(self).union(*seqs)
+
+    def intersection(self, *seqs):
+        return frozenset(self).intersection(*seqs)
+
+    def difference(self, *seqs):
+        return frozenset(self).difference(*seqs)
+
+    def symmetric_difference(self, *seqs):
+        return frozenset(self).symmetric_difference(*seqs)
+
+    def __eq__(self, seq):
+        if not isinstance(seq, (set, frozenset, proxied_frozenset, mapped_frozenset)) or (
+              len(self) != len(seq)):
+            return False
+        for val in self:
```

Comment:
Don't use naive iteration, it's slow. Reach for the buffer and iterate the buffer in an optimized fashion.

Reply:
Is it really that slow ? proxied_list's ```_getitem``` already fetches the buffer and dereferences the pointer. The only source of overhead I can think of is the tuple that is created by the ```_metadata``` method, but we can work around that.

Reply:
`for val in self` has to construct and iterate a generator.

Yes, it's very high overhead, even if we optimize the generator. At the very least you should inline the generator loop to avoid the generator indirection.

Reply:
Good point. A C-style index iteration should be much faster. Thanks.